### ISSUE TITLE
COM Registration Fixing tool doesn't work

### ISSUE THREAD

**Issue by tspivey** [nvaccess/nvda#9039]
### Steps to reproduce:
1. Install MathPlayer.
2. Uninstall MathPlayer.
3. Read the task bar and notice it says unknown.
4. Run the COM Registration Fixing tool.
5. Read the task bar again and it still says unknown.

### Actual behavior:
The tool doesn't fix NVDA's accessibility support.

### Expected behavior:
The tool fixes NVDA's accessibility support.

#### NVDA installed/portable/running from source:
Installed

#### NVDA version:
Version: alpha-16405,19bb21c2

#### Windows version:
Version 1809 (OS Build 17763.134)

#### Does the issue still occur after restarting your PC?
Yes

#### Have you tried any other versions of NVDA?
No

**Comment by leonardder** [comment]
What happens if you execute all the tweaks in https://github.com/nvaccess/nvda/issues/2807#issuecomment-320149243 by hand?

**Comment by Adriani90** [comment]
cc: @jcsteh, @michaelDCurran

**Comment by tspivey** [comment]
There are a lot of them. I haven't tried all of them, but here's a summary of some things I did try.

Testing with a win7 VM, this fixes it, note that this doesn't work on my real win10 machine, and requires an NVDA restart:
    c:\windows\syswow64\regsvr32 c:\windows\syswow64\oleacc.dll

On my win10 machine, this python console command fixes it if NVDA is run as administrator, but I forgot exactly what it does. It seems to return an error, but I'm not sure how to turn the int back into a HRESULT. However, it doesn't work on the win7 VM:
    import comtypes,ctypes;p=comtypes.typeinfo.LoadTypeLibEx(r"c:\windows\syswow64\oleacc.dll");ctypes.windll.oleaut32.RegisterTypeLib(p, ur"c:\windows\syswow64\oleacc.dll", None)

On the win10 machine, Registering c:\windows\syswow64\oleacc.dll and c:\windows\syswow64\oleaut32.dll didn't fix it.
I also tried the ones in system32, along with actxprxy.dll. No luck.

**Comment by leonardder** [comment]
Just to make sure, did you try OneCoreCommonProxyStub.dll on Windows 10 
as well?

**Comment by tspivey** [comment]
@leonardder  I didn't realize you linked to a comment, let's try these.

0. Install/uninstall MathPlayer.
1. First reg file - now my start button doesn't say unknown, but that might've been because of the reboot since I didn't test beforehand.
2. Second reg file - no luck.
3. Started an admin cmd. Hey, my Firefox with the instructions still reads, maybe I'm in different bug territory now.
4. Entered both commands, restarted NVDA. No task bar yet.
5. Did that, no taskbar yet.
6. I'm on win10, and this module isn't there.
7. Same as 6.
8. The module "c:\windows\system32\OneCoreCommonProxyStub.dll" was loaded but the entry-point DllRegisterServer was not found.
9. The module "c:\windows\syswow64\OneCoreCommonProxyStub.dll" was loaded but the entry-point DllRegisterServer was not found.

Next steps:
1. Run COM registration fixing tool, nothing, I figured it would at least get it back to saying unknown.
2. Magic python console incantation - got my system back.
3. Don't bother with MathPlayer because I'd have to reboot again and just run the first reg file, my start button doesn't say unknown anymore. I guess that breaks it rather than fixes it.
4. Python console again - back to a working system.

**Comment by leonardder** [comment]
Hmm, so oleaut32.RegisterTypeLib does work why registering it with resvr32 does not, is that correct?

**Comment by tspivey** [comment]
Yes, at least on win10.

**Comment by jcsteh** [comment]
I wonder whether IDispatch still needs the typelib registered correctly, even if oleacc.dll is used as the marshaler. I guess that would make some sense. NVDA shouldn't be using IDispatch directly, but it could still feasibly be causing weirdness.

**Comment by michaelDCurran** [comment]
@jcsteh do you know if there are separate typelib registrations required 
for 64 and 32 bit?

Or is running it from within NVDA (32 bit) enough?

**Comment by jcsteh** [comment]
I'm not certain. [This StackOverflow thread](https://stackoverflow.com/questions/9914535/what-decides-whether-registertypelib-writes-to-win32-or-win64) suggests they can be different. On the flip side, that same thread also gives me the impression that you can register a 64 bit typelib even in a 32 bit process.

**Comment by michaelDCurran** [comment]
@tspivey: I want to get this into the COM registration fixing tool for 2019.1, however, with my copy of mathplayer that I have, on my system uninstalling it does not cause any of the problems you mention, though I do admit I have seen the problem a few years back. I thought that Design Science had addressed the issue in their installer.
If you could share your specific copy of the mathplayer installer with me, so that I can reproduce the issue, that would be helpful.

**Comment by tspivey** [comment]
https://www.dropbox.com/s/mxq3nwo6th4ln15/MathPlayerSetup.exe?dl=1

**Comment by lukaszgo1** [comment]
@michaelDCurran Have you made any progress here?

**Comment by Adriani90** [comment]
@tspivey does this occur only in case of mathplayer? Or did you see this behavior with installing and uninstalling other programs as well?
Just for the reference: which programs do you have installed on your PC? Is there something like skype for business for example?
Skype for business is also causing some issues related to unregistered COM DLLs.

**Comment by tspivey** [comment]
No, I haven't personally seen it with anything other than mathplayer. But since I've had to help people fix it, it's still a problem.

**Comment by josephsl** [comment]
Hi,

Update: we had a user report that things in web browsers were not read, and the workaround noted by Tyler did work. I think, for 2020.1, the workaround line should be attempted unless we can fine a more refined solution for 2020.2.

Thanks.

**Comment by cary-rowen** [comment]
This is still confusing, and I can still reproduce this issue using 2020.4.
Is there any news about this?
@michaelDCurran
@josephsl

**Comment by jcsteh** [comment]
Aside from the typelib issue here, it seems the COM registration tool isn't applying the registry patch correctly, but I don't know why.

STR (on a 64 bit system):

1. Press Windows+r, type regedit, press enter.
2. In the address bar, enter:
    `Computer\HKEY_CLASSES_ROOT\Interface\{618736E0-3C3D-11CF-810C-00AA00389B71}\ProxyStubClsid32`
3. Change the (Default) value to:
    `{00020424-0000-0000-C000-000000000046}`
4. NVDA menu -> Tools -> Run COM Registration Fixing tool. Accept the warning, allow administrator access, press OK.
5. Refresh registry editor by pressing f5.
    - Expected: The (Default) value should be `{03022430-ABC4-11D0-BDE2-00AA001A1953}` as specified in the NVDA oleaccProxy.reg patch.
    - Actual: The (Default) value is `{00020424-0000-0000-C000-000000000046}`

What's odd is that if I rename the .reg file, an error gets logged, so the function is definitely running. And if I manually run:
`c:\windows\regedit.exe /s "C:\Program Files (x86)\NVDA\COMRegistrationFixes\oleaccProxy.reg"`
it updates the registry as expected.

**Comment by XLTechie** [comment]
Trying to figure this out is part of my plan for #12355, along with better 
error reporting and some other things requested by some people during my 
previous attempt to refactor the CRFT.

That said, if someone solves it sooner, that would be welcome. :)

**Comment by jcsteh** [comment]
Oh! I think I understand what's going wrong here. In 32 bit apps, Windows WOW64 [redirects certain file system paths](https://docs.microsoft.com/en-us/windows/win32/winprog64/file-system-redirector). So, when you run c:\windows\regedit.exe, this gets redirected transparently to c:\windows\syswow64\regedit.exe. The same happens for regsvr32. This means this tool has never worked for the 64 bit registry, only the 32 bit registry.

WOW64 recognises c:\windows\sysnative as a special path which redirects to the native (64 bit) system directory. However, regedit.exe isn't in system, so this may not work for that. If not, we'll have to disable WOW64 file system redirection before our subprocess calls and re-enable it again afterward.

**Comment by jcsteh** [comment]
[This thread](https://social.msdn.microsoft.com/Forums/sqlserver/en-US/190d5840-5861-43fc-b95d-ff0e702529bc/how-to-open-64-bit-regeditexe-from-32-bit-application?forum=vbgeneral) confirms that indeed you can't do sysnative\regedit.exe. However, you can apparently do sysnative\regedt32.exe, which apparently redirects to the 64 bit regedit. What a mess.

**Comment by jooyoungseo** [comment]
NVDA keeps saying "unknown" in my Firefox, and the "Run COM Registration Fixing tool" does not take any effect. I use Windows 10 64-bit. I know this issue was mentioned in another thread and the solutions were suggested; however, would you mind sharing step-by-step instructions here for Windows 10 64-bit users? I tried to fix the issue on my end following the suggested comments in other threads, but none of them worked.

**Comment by jcsteh** [comment]
Is this recent; i.e. only with the Firefox 89 update? If so, this is due to
a bug in Firefox. Please update to Firefox 89.0.1, which should fix this.

If this isn't recent (i.e. it happens with older versions of Firefox too),
this is due to COM registration issues.

**Comment by jooyoungseo** [comment]
@jcsteh, mine was 89, so have just updated to 89.0.1 following your suggestion. Still, I have the same issue. I also tried with the COM fixing tool and did reboot. No success.

**Comment by jcsteh** [comment]
Please try the following:
1. Press Windows+r to open the run dialog.
2. Paste the following:
%ProgramFiles(x86)%\nvda\COMRegistrationFixes\oleaccProxy.reg
3. Allow admin access, then allow registry editor to make the changes.
4. Reboot.

**Comment by jooyoungseo** [comment]
Unfortunately, no success. Still, "unknown".

**Comment by jcsteh** [comment]
Okay. Try this (very similar, but a slightly different command):
1. Press Windows+r to open the run dialog.
2. Paste the following:
c:\windows\syswow64\regedit.exe
%ProgramFiles(x86)%\nvda\COMRegistrationFixes\oleaccProxy.reg
3. Allow admin access, then allow registry editor to make the changes.
4. Reboot.

**Comment by jooyoungseo** [comment]
I see the following error three times:

> Registry Editor
> Cannot import C:\Program: Error opening the file. There may be a disk or file system error.
> OK

**Comment by jcsteh** [comment]
My bad. Try this (note the quotes):
c:\windows\syswow64\regedit.exe
"%ProgramFiles(x86)%\nvda\COMRegistrationFixes\oleaccProxy.reg"

**Comment by jooyoungseo** [comment]
With quotes, it was registered, but still it says "unknown" in Firefox after reboot.

**Comment by XLTechie** [comment]
> With quotes, it was registered, but still it says "unknown" in Firefox after reboot.

Same thing found by a user on the NVDA list to whom I gave similar 
instructions last week. Something is still missing. I don't believe it was 
Firefox in that case, although I'd have to go back to verify.

**Comment by MichelSuch** [comment]
Hi,


Just a thought:

Remember that Firefox somewhere has a checkbox to enable/disable 
accessibilité API.

Could it have been unchecked for some obscure reason?




Michel Such ***@***.***>
------------------------------------------------------------------------
Expert NVDA certifié 2019
Le 17/06/2021 à 07:41, Luke Davis a écrit :

**Comment by jcsteh** [comment]
Very strange. It could be the typelib as discussed earlier in this issue,
but I need to dig up the right registry keys to fix that, as calling the
function from a 32 bit executable may not cover the 64 bit case; I'm not
sure.

**Comment by josephsl** [comment]
Hi, you may need to specify 64-bit key access (I found out the hard way while working on Resource Monitor a few years ago when obtaining UBR (update build revision) for a given Windows 10 build). Thanks.

**Comment by jcsteh** [comment]
We don't need to do that in this case because we apply the patch using the
two different versions of regedit.exe... although as I noted above, we're
not actually doing what we think we're doing there. The fix is relatively
straightforward, but I don't have a working dev environment since NVDA
moved back to Python 3.7 and I haven't had time to rectify that.

**Comment by josephsl** [comment]
Hi, I think a diff file would suffice in letting us test it in our environments, or perhaps a pseudo-code so one of us can commit it on your behalf? Thanks.

**Comment by XLTechie** [comment]
@jcsteh I have done a fix for this in #12560 .
I have changed the executable to `regedt32.exe`, and I'm sorting out the paths.
However:

Going back to the original code. If you assume the correct executable is being used, I still think this doesn't work as intended.

In `COMRegistrationFixes/__init__.py`, there is a test for whether this is 64 bit Windows. If it is, it sets wow64 True on the `applyRegistryPatch` function call.
I think that has confused the issue, maybe leading to these funky paths.

I think what this function wants to know, is whether it is running under 64 bit Windows, **and** whether we want a 32bit or a 64bit registry operation.
Currently, it tries to combine those questions into one value, and I think that has lead to this.

If I am correct, what we need to provide is:
* Bool: is this 64bit Windows.
* Bool: do we want a 64bit registry operation.

1. If it is a 32bit operation, just call `%WINDIR%\System32\regedt32.exe`. Either this is a 32bit system, and System32 is correct; or SysWOW64 will worry about redirecting to the proper executable, which is its purpose. We shouldn't need to care.
2. If this is a 64bit operation, it should check whether is64bit is set True. If it is, we really do need the alternative path to reach the 64bit editor. In this case, that path is `%WINDIR%\Sysnative\regedt32.exe`, so we call that.

In no case should we ever call anything in `%WINDIR%\SysWOW64` ourselves, which we currently do.

Have I missed anything here? If not, I will alter the PR accordingly.

**Comment by XLTechie** [comment]
@jcsteh, @josephsl: pursuant to my previous comment, I went ahead and 
updated #12560 on the assumption that I am correct.

I created separate functions for applying 32-bit and 64-bit registry 
patches, just to make it clear what's going on when their called. That 
seemed cleaner than a flag to control the direction of the process.

As a separate issue, please also look at the DLL registrations.
Some of them are being taken from the SysWOW64 directory after a 64-bit 
check, which makes me think that 64-bit libraries are expected. However, 
if I'm not mistaken, 32-bit libraries are all we'll get from that folder.

Before I finalize the PR (Reef said he might get it in beta), I need to 
know if the DLL registrations need to be worked on.

**Comment by jooyoungseo** [comment]
@XLTechie, do you have any suggestions for me to fix the Firefox issue until the PR is fully tested and merged? I would like to address this asap if possible. Or, is my issue subject to further investigation?

**Comment by XLTechie** [comment]
@jooyoungseo here is a [try build](https://ci.appveyor.com/api/buildjobs/1dp10lngn3a5n8ai/artifacts/output%2Fnvda_snapshot_pr12560-23130%2C588bd2cc.exe), if you are feeling adventurous. Please let us know the result.
I suggest rebooting after running the COM Registration Fix.

**Comment by XLTechie** [comment]
@jcsteh I have tested the [try build](https://ci.appveyor.com/api/buildjobs/1dp10lngn3a5n8ai/artifacts/output%2Fnvda_snapshot_pr12560-23130%2C588bd2cc.exe) from #12560.

I have found that it performs correctly given your ProxyStubClsid32 test in the STR above. However, at least on my system, NVDA becomes somewhat unstable after this, and requires a restart.

I am unclear whether that is from munging the ProxyStubClsid32 key and then changing it back; from doing this with an unsigned executable running in admin mode (shouldn't be that); or because something else is going on.
Can you please test your own STR and this try build, and see what kind of results you get?
@josephsl if you have time to do the same, that would be appreciated.

If this does make NVDA unstable, but is otherwise working correctly, we may need to add a restart warning to the dialogs.

**Comment by jooyoungseo** [comment]
@XLTechie -- Thanks for your help. I executed the installer with admin Privileged, and hit the "Run Anyway" button.

After hearing the start-up chime sound from the installer, I got the following error:

> explorer.exe - Application Error
> explorer.exe - Application Error
> The application was unable to start correctly (0xc0000142). 
> Click OK to close the application. 
> OK 

I clicked the OK, and then the following error popped up:

> csrss.exe - Application Error
> csrss.exe - Application Error
> The application was unable to start correctly (0xc0000142). 
> Click OK to close the application. 
> OK

**Comment by XLTechie** [comment]
Are you on a machine with a touchscreen?

**Comment by jooyoungseo** [comment]
No

**Comment by XLTechie** [comment]
@jcsteh or @josephsl  Can I ask why we even need to do the 32-bit and 64-bit registry edits?

Quoting from Microsoft's [wow64_bestprac.docx](https://download.microsoft.com/download/a/f/7/af7777e5-7dcd-4800-8a0a-b18336565f5b/wow64_bestprac.docx) document, which describes the different views that 32 and 64 bit programs get of the registry:
> Registry Reflection
> In versions of Windows earlier than Windows 7, some redirected registry keys were reflected as this section describes. In Windows 7 and Windows Server 2008 R2, no keys are reflected. Formerly reflected keys are now shared and are visible to both the 32-bit and 64-bit registry hives. For more information, see “Removal of Windows Registry Reflection” on the MSDN website.

Since we only support Win 7 and above, is there any value in continuing to do this?

**Comment by XLTechie** [comment]
@jooyoungseo Please try [this try-build](https://ci.appveyor.com/api/buildjobs/kqwv5s0f19vgy0ao/artifacts/output%2Fnvda_snapshot_pr12560-23151%2C4b6b77b5.exe)

**Comment by XLTechie** [comment]
@tspivey, @cary-rowen.
Can you two replicate any of the problems that you had, and can you test the most recent try-build of #12560, to see if it can rectify them?
If you do, please post here your results, along with the bitness of your system, and Windows version (7, 10).
I need to know if more needs to be done there.
Thank you.

**Comment by cary-rowen** [comment]
Hello, this is the log I executed on Win1021H1. It has a dll registration error.
[Nvda-log.txt](https://github.com/nvaccess/nvda/files/6693152/Nvda-log.txt)
@XLTechie

**Comment by XLTechie** [comment]
@leonardder @jcsteh 
I decided to try @tspivey's mathPlayer test from the start of this thread.
I installed his buggy MathPlayer, and then uninstalled it.
The result, is that my Firefox, upon restart, was left saying "unknown" for anything that was actually part of a page, though the Firefox elements themselves work fine (menues, address bar).
I switched over to Narrator, and ran the try-build from #12560 as a temporary copy, and the CRFT it contains.
No errors reported from the CRFT run, but no fix of the "unknown", either. I did seem to get my taskbar back at that step, I believe.

Next, I switched back to latest Alpha, and tried running @tspivey's [typeLib code](https://github.com/nvaccess/nvda/issues/9039#issuecomment-445803675) (modified to work in py3, and also modified to let the system choose the dll it wanted (I used system32 as the path not syswow64)).
The end result was the following return code:
`-2147319780`
And no apparent fix.
I tried again with his original paths, but still no joy. Lastly i tried with Sysnative (the 64-bit versions), and got nothing but the same int returned.

-----

I have to say I am at a bit of a loss at this point. @michaelDCurran I don't think this is the fault of the PR, I think it still does a better job than the old tool. I tested the old tool in alpha, and it did no better. So I don't *think* this should hold up #12560, but there's obviously more to do.

-----

While in Firefox, here is the dev info for an "unknown":
```

Developer info for navigator object:
name: None
role: ROLE_UNKNOWN
roleText: None
states: 
isFocusable: False
hasFocus: False
Python object: <NVDAObjects.IAccessible.IAccessible object at 0x08B21BD0>
Python class mro: (<class 'NVDAObjects.IAccessible.IAccessible'>, <class 'NVDAObjects.window.Window'>, <class 'NVDAObjects.NVDAObject'>, <class 'documentBase.TextContainerObject'>, <class 'baseObject.ScriptableObject'>, <class 'baseObject.AutoPropertyObject'>, <class 'garbageHandler.TrackedObject'>, <class 'object'>)
description: None
location: None
value: None
appModule: <'appModuleHandler' (appName 'firefox', process ID 9460) at address 6fa80d0>
appModule.productName: 'Firefox'
appModule.productVersion: '89.0.1'
TextInfo: <class 'NVDAObjects.NVDAObjectTextInfo'>
windowHandle: 852830
windowClassName: 'MozillaWindowClass'
windowControlID: 0
windowStyle: 399441920
extendedWindowStyle: 256
windowThreadID: 3468
windowText: 'Refactored the COM Registration Fixing Tool code to remove bugs and for better UX by XLTechie · Pull Request #12355 · nvaccess/nvda — Mozilla Firefox'
displayText: exception: cannot unpack non-iterable NoneType object
IAccessibleObject: <POINTER(IAccessible) ptr=0xa6fd8a0 at 8af46c0>
IAccessibleChildID: -117441110
IAccessible event parameters: windowHandle=852830, objectID=-4, childID=-117441110
IAccessible accName: exception: (-2147024809, 'The parameter is incorrect.', (None, None, None, 0, None))
IAccessible accRole: exception: (-2147024809, 'The parameter is incorrect.', (None, None, None, 0, None))
IAccessible accState: exception: (-2147024809, 'The parameter is incorrect.', (None, None, None, 0, None))
IAccessible accDescription: exception: (-2147024809, 'The parameter is incorrect.', (None, None, None, 0, None))
IAccessible accValue: exception: (-2147024809, 'The parameter is incorrect.', (None, None, None, 0, None))
```

Lastly, I have a desktop shortcut key setup to taskkill Firefox.
While in Firefox, if I press it, a COM NULL pointer is accessed:

```

IO - speech.speech.speak (05:40:37.369) - MainThread (12456):
Speaking ['Taskbar', CancellableSpeech (still valid)]
IO - speech.speech.speak (05:40:37.529) - MainThread (12456):
Speaking ['Firefox killer', 'window', CancellableSpeech (still valid)]
DEBUGWARNING - NVDAObjects.__call__ (05:40:37.578) - Dummy-1 (11528):
Invalid NVDAObject: no windowHandle
Traceback (most recent call last):
  File "NVDAObjects\__init__.pyc", line 83, in __call__
  File "NVDAObjects\UIA\__init__.pyc", line 1122, in __init__
NVDAObjects.InvalidNVDAObject: no windowHandle
DEBUGWARNING - NVDAObjects.__call__ (05:40:37.580) - Dummy-1 (11528):
Invalid NVDAObject: no windowHandle
Traceback (most recent call last):
  File "NVDAObjects\__init__.pyc", line 83, in __call__
  File "NVDAObjects\UIA\__init__.pyc", line 1122, in __init__
NVDAObjects.InvalidNVDAObject: no windowHandle
DEBUGWARNING - NVDAObjects.__call__ (05:40:37.583) - Dummy-1 (11528):
Invalid NVDAObject: no windowHandle
Traceback (most recent call last):
  File "NVDAObjects\__init__.pyc", line 83, in __call__
  File "NVDAObjects\UIA\__init__.pyc", line 1122, in __init__
NVDAObjects.InvalidNVDAObject: no windowHandle
DEBUGWARNING - NVDAObjects.__call__ (05:40:37.585) - Dummy-1 (11528):
Invalid NVDAObject: no windowHandle
Traceback (most recent call last):
  File "NVDAObjects\__init__.pyc", line 83, in __call__
  File "NVDAObjects\UIA\__init__.pyc", line 1122, in __init__
NVDAObjects.InvalidNVDAObject: no windowHandle
DEBUGWARNING - NVDAObjects.__call__ (05:40:37.588) - Dummy-1 (11528):
Invalid NVDAObject: no windowHandle
Traceback (most recent call last):
  File "NVDAObjects\__init__.pyc", line 83, in __call__
  File "NVDAObjects\UIA\__init__.pyc", line 1122, in __init__
NVDAObjects.InvalidNVDAObject: no windowHandle
DEBUGWARNING - NVDAObjects.__call__ (05:40:37.590) - Dummy-1 (11528):
Invalid NVDAObject: no windowHandle
Traceback (most recent call last):
  File "NVDAObjects\__init__.pyc", line 83, in __call__
  File "NVDAObjects\UIA\__init__.pyc", line 1122, in __init__
NVDAObjects.InvalidNVDAObject: no windowHandle
DEBUGWARNING - NVDAObjects.__call__ (05:40:37.593) - Dummy-1 (11528):
Invalid NVDAObject: no windowHandle
Traceback (most recent call last):
  File "NVDAObjects\__init__.pyc", line 83, in __call__
  File "NVDAObjects\UIA\__init__.pyc", line 1122, in __init__
NVDAObjects.InvalidNVDAObject: no windowHandle
DEBUGWARNING - NVDAObjects.__call__ (05:40:37.594) - Dummy-1 (11528):
Invalid NVDAObject: no windowHandle
Traceback (most recent call last):
  File "NVDAObjects\__init__.pyc", line 83, in __call__
  File "NVDAObjects\UIA\__init__.pyc", line 1122, in __init__
NVDAObjects.InvalidNVDAObject: no windowHandle
DEBUGWARNING - NVDAObjects.__call__ (05:40:37.596) - Dummy-1 (11528):
Invalid NVDAObject: no windowHandle
Traceback (most recent call last):
  File "NVDAObjects\__init__.pyc", line 83, in __call__
  File "NVDAObjects\UIA\__init__.pyc", line 1122, in __init__
NVDAObjects.InvalidNVDAObject: no windowHandle
DEBUGWARNING - NVDAObjects.__call__ (05:40:37.597) - Dummy-1 (11528):
Invalid NVDAObject: no windowHandle
Traceback (most recent call last):
  File "NVDAObjects\__init__.pyc", line 83, in __call__
  File "NVDAObjects\UIA\__init__.pyc", line 1122, in __init__
NVDAObjects.InvalidNVDAObject: no windowHandle
ERROR - UIAUtils._isImprovedConhostTextRangeAvailable (05:40:37.607) - MainThread (12456):
Traceback (most recent call last):
  File "UIAUtils.pyc", line 329, in _isImprovedConhostTextRangeAvailable
  File "comtypesMonkeyPatches.pyc", line 27, in __call__
ValueError: NULL COM pointer access
DEBUGWARNING - NVDAObjects.UIA.UIA._prefetchUIACacheForPropertyIDs (05:40:37.617) - MainThread (12456):
IUIAutomationElement.buildUpdatedCache failed given IDs of {30019, 30086, 30022, 30025, 30155, 30036, 30070, 30103, 30008, 30009, 30010, 30046, 30079}
DEBUGWARNING - eventHandler.executeEvent (05:40:37.623) - MainThread (12456):
error executing event: gainFocus on <NVDAObjects.Dynamic_WinConsoleUIAEditableTextWithAutoSelectDetectionUIA object at 0x08AFF270> with extra args of {}
Traceback (most recent call last):
  File "eventHandler.pyc", line 246, in executeEvent
  File "eventHandler.pyc", line 96, in __init__
  File "eventHandler.pyc", line 105, in next
  File "NVDAObjects\behaviors.pyc", line 354, in event_gainFocus
  File "NVDAObjects\behaviors.pyc", line 187, in event_gainFocus
  File "NVDAObjects\UIA\__init__.pyc", line 1141, in event_gainFocus
  File "NVDAObjects\__init__.pyc", line 1137, in event_gainFocus
  File "NVDAObjects\__init__.pyc", line 1005, in reportFocus
  File "speech\speech.pyc", line 538, in speakObject
  File "speech\speech.pyc", line 568, in getObjectSpeech
  File "baseObject.pyc", line 42, in __get__
  File "baseObject.pyc", line 146, in _getPropertyViaCache
  File "NVDAObjects\__init__.pyc", line 1368, in _get__hasNavigableText
  File "baseObject.pyc", line 42, in __get__
  File "baseObject.pyc", line 146, in _getPropertyViaCache
  File "NVDAObjects\UIA\__init__.pyc", line 1421, in _get_states
  File "NVDAObjects\UIA\__init__.pyc", line 838, in _getUIACacheablePropertyValue
  File "comtypesMonkeyPatches.pyc", line 27, in __call__
_ctypes.COMError: (-2147220991, 'An event was unable to invoke any of the subscribers', (None, None, None, 0, None))
DEBUG - appModuleHandler.cleanup (05:40:37.649) - MainThread (12456):
application firefox closed
DEBUG - appModuleHandler.cleanup (05:40:37.649) - MainThread (12456):
application taskkill closed
DEBUG - appModuleHandler.cleanup (05:40:37.649) - MainThread (12456):
application conhost closed
IO - speech.speech.speak (05:40:37.657) - MainThread (12456):
Speaking ['Taskbar', CancellableSpeech (still valid)]
IO - inputCore.InputManager.executeGesture (05:40:40.135) - winInputHook (5204):
```

**Comment by XLTechie** [comment]
@cary-rowen It looks like you downloaded an earlier build. Try [this link](https://ci.appveyor.com/api/buildjobs/kqwv5s0f19vgy0ao/artifacts/output%2Fnvda_snapshot_pr12560-23151%2C4b6b77b5.exe).

**Comment by jooyoungseo** [comment]
@XLTechie, no luck -- still having "unknown" controls in Firefox instead of actual content although the new test build has been successfully installed. The "COM Registration Fixing tool" still does not work.

## Microsoft Windows Version

> Version 20H2 (OS Build 19042.1052) 64 bit

**Comment by michaelDCurran** [comment]
@XLTechie  On Windows 10 64 bit build  21390 I installed and then uninstalled the Mapthplayer linked in this issue. As expected my Firefox started saying unknown everywhere. I then ran the COM registration fixing tool from the latest try build of your branch, and restarted NvDA, and Firefox started working perfectly again.
Are you using Firefox 32 bit or 64 bit? Did you restart NVDA and or your system after running the COM registration fixing tool?

**Comment by michaelDCurran** [comment]
No sorry I spoke too soon. It fixed the Library (download history) Window, and any existing open tabs still work, but all new tabs and windows are broken. However, Chrome is not broken, thus I think it may have something to do with Firefox's inter-process proxying code. There may be another interface from the MSAA typelibrary we need to correct its COM marshaller for.

**Comment by jooyoungseo** [comment]
FYI, of course, I restarted my system after the NVDA snapshot installation. Plus, rebooted after applying the "COM Registration Fixing tool." I am using 64 bit Firefox (v89.0.1). No success.

**Comment by Mary5958** [comment]
Does it work in troubleshoot mode (previously safe mode)? If so then try changing the theme in Firefox to something else than the default one. I did to Firefox Alpenglow and got it working still in version 89.0 this way, but in 89.01 it already works fine here with default settings. Maybe completely uninstalling Firefox along with related data and installing again would also help.

**Comment by ABuffEr** [comment]
> It fixed the Library (download history) Window, and any existing open tabs still work, but all new tabs and windows are broken. However, Chrome is not broken, thus I think it may have something to do with Firefox's inter-process proxying code.

I saw exactly the same situation over a PC of a (not advanced) user I assisted remotely, after running various fixes manually. I'm following discussion with great interest.

**Comment by XLTechie** [comment]
@michaelDCurran
I have installed mathPlayer again, which of course fixes the Firefox problem.
Here is a registry diff between non-working (before install) and working (after install) states.
[registry diff.txt](https://github.com/nvaccess/nvda/files/6707355/registry.diff.txt)

I am not sure what I'm looking for here. I have tried variations of this on a couple systems, including a diff between a working system, then an install and remove of MathPlayer. Unfortunately, Firefox updated and Windows Security updated, while I was doing it, even though I had WIFI off until near the end to test Firefox.
As a result, the diff there had over a hundred thousand lines listed, between added, removed, and changed keys.
I didn't think that was useful, so went back to this original system, and got the diff above.

**Comment by XLTechie** [comment]
@Mary5958 Unfortunately, there are other apps which are effected by this. For example, regshot, the program I used to make that registry diff, has menus which stop working after MathPlayer is uninstalled. At least on one system, Notepad wouldn't read properly.

**Comment by jooyoungseo** [comment]
Very interestingly, my Firefox works again after installing MathPlayer.

**Comment by XLTechie** [comment]
> Very interestingly, my Firefox works again after installing MathPlayer.

Yes. MathPlayer is very aggressive in its uninstallation, and takes 
something with it that it shouldn't.
When it is installed, it doesn't assume that something to be there, so it 
installs it, thus resolving the problem.

Trying to figure out what that something is (presumably a registry key 
that registers a particular interface) is what is underway now.

**Comment by michaelDCurran** [comment]
Here is another try build for people to test: https://ci.appveyor.com/api/buildjobs/04lnhv7snv1fe12l/artifacts/output%2Fnvda_snapshot_try-oleaccProxyFix-23189%2C2ac08400.exe
Please install the broken MathPlayer. Then uninstall it. Confirm that Mozilla Firefox now only says unknown in its documents. 
then run the COM registration fixing tool in this try build and restart NVDa / firefox, or better, reboot.
Then confirm that Firefox again works.
 
This is based on pr #12560 but adds an extra change to the oleaccProxy.reg file, namely that it deletes the TypeLib subkey from the IAccessible interface key. This was never initially done as that file changed the proxyClsid32 value to point directly to oleacc.dll rather than the automation typelib marshaller, thus Windows would never have a reason to look at the TypeLib subkey (whether it was correct or not). But it seems that Firefox specifically looks at this subkey or at very least tries to get the IAccessible type library and fails, as it points to the removed Mathplayer type library.

I'm mostly interested in this being tested on windows 7. I have confirmed this works okay on Windows 10.

**Comment by XLTechie** [comment]
@michaelDCurran a couple things.

1. It didn't happen the first time, but after I removed MathPlayer the second time from Win 10 20H2, it took out accessibility for Brave as well as Firefox.
With that in mind, I checked on the 1909 system I have, and Chrome was also giving me "unknowns", after only one removal of MathPlayer.

2. The tool works with this change in Windows 10 as you said, but only after a reboot. I think I should add a "please reboot now" emphatic warning to the exit dialog.

3. Should we verify the value of this key is bad before just blowing it away?

4. When the tool runs, it dumps this in the log. You probably got this yourself, but just in case:

```
Speaking [LangChangeCommand ('en_GB'), 'COM Registration Fixing Tool', 'dialog', 'COM Registration Fixing tool complete', CancellableSpeech (still valid)]
IO - speech.speech.speak (01:01:47.971) - MainThread (14332):
Speaking [LangChangeCommand ('en_GB'), 'OK', 'button', CancellableSpeech (still valid)]
DEBUGWARNING - winConsoleHandler.connectConsole (01:01:49.291) - MainThread (14332):
Could not attach console: OSError(9, 'The handle is invalid.', None, 6)
DEBUGWARNING - winConsoleHandler.connectConsole (01:01:49.623) - MainThread (14332):
Could not attach console: OSError(22, 'The server is currently disabled.', None, 1341)
IO - speech.speech.speak (01:01:49.633) - MainThread (14332):
Speaking [LangChangeCommand ('en_GB'), 'terminal', CancellableSpeech (still valid)]
ERROR - NVDAObjects.behaviors.LiveText._monitor (01:01:49.365) - EnhancedLegacyWinConsole._monitorThread (12664):
Error getting initial text
Traceback (most recent call last):
  File "NVDAObjects\behaviors.pyc", line 310, in _monitor
  File "NVDAObjects\window\winConsole.pyc", line 66, in _getText
  File "winConsoleHandler.pyc", line 127, in getConsoleVisibleLines
  File "wincon.pyc", line 77, in GetConsoleScreenBufferInfo
OSError: [WinError 6] The handle is invalid.
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (01:01:49.862) - MainThread (14332):
accRole failed: (-2147024809, 'The parameter is incorrect.', (None, None, None, 0, None))
IO - speech.speech.speak (01:01:50.108) - MainThread (14332):
Speaking [LangChangeCommand ('en_GB'), 'COM Registration Fixing Tool', 'dialog', 'COM Registration Fixing tool complete', CancellableSpeech (still valid)]
IO - speech.speech.speak (01:01:50.116) - MainThread (14332):
Speaking [LangChangeCommand ('en_GB'), 'OK', 'button', CancellableSpeech (still valid)]
DEBUGWARNING - winConsoleHandler.connectConsole (01:01:50.282) - MainThread (14332):
Could not attach console: PermissionError(13, 'Access is denied.', None, 5)
DEBUGWARNING - winConsoleHandler.connectConsole (01:01:50.332) - MainThread (14332):
Could not attach console: PermissionError(13, 'Access is denied.', None, 5)
IO - speech.speech.speak (01:01:50.341) - MainThread (14332):
Speaking [LangChangeCommand ('en_GB'), 'C:\\WINDOWS\\Sysnative\\reg.exe', 'terminal', CancellableSpeech (still valid)]
ERROR - NVDAObjects.behaviors.LiveText._monitor (01:01:50.314) - EnhancedLegacyWinConsole._monitorThread (13904):
Error getting initial text
Traceback (most recent call last):
  File "NVDAObjects\behaviors.pyc", line 310, in _monitor
  File "NVDAObjects\window\winConsole.pyc", line 66, in _getText
  File "winConsoleHandler.pyc", line 127, in getConsoleVisibleLines
  File "wincon.pyc", line 77, in GetConsoleScreenBufferInfo
OSError: [WinError 6] The handle is invalid.
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (01:01:50.553) - MainThread (14332):
accRole failed: (-2147024809, 'The parameter is incorrect.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (01:01:50.632) - MainThread (14332):
accRole failed: (-2147024809, 'The parameter is incorrect.', (None, None, None, 0, None))
IO - speech.speech.speak (01:01:50.831) - MainThread (14332):
Speaking [LangChangeCommand ('en_GB'), 'COM Registration Fixing Tool', 'dialog', 'COM Registration Fixing tool complete', CancellableSpeech (still valid)]
IO - speech.speech.speak (01:01:50.842) - MainThread (14332):
```

**Comment by michaelDCurran** [comment]
That TypeLib key does not exist at all on default installs of Windows 
10, and I'm assuming Windows 7 either, and is only put there by 
MathPlayer, as the IAccessible interface is incorrectly included in 
their type library, and therefore registered like that when mathPlayer 
is installed.

So, I don't believe there really is a valid value for that key anyway.
